File: infrastructure/mcp/server/src/resources.ts
// SPDX-License-Identifier: MIT
//
// Resource resolvers — one per URI kind. Each resolver reads what it
// needs from the Universe filesystem and returns a `{uri, mimeType,
// text}` block fit for `resources/read`.
//
// Resolvers are intentionally separate from tools. Tools are actions
// the AI invokes; resources are memory the AI pins and refers back to.
// The split means a tool like `eustress_read_script` and a resource at
// `eustress://script/...` can diverge in formatting (tools prefer
// JSON, resources prefer human-readable markdown) without either
// duplicating logic for the other.

import * as fs from 'node:fs/promises';
import * as fsSync from 'node:fs';
import * as path from 'node:path';
import {
  listSpaces, listScripts, readCapped, extractToml,
  MAX_FILE_BYTES,
} from './universe.js';
import { buildUri, parseUri, type EustressUri } from './uri.js';

export type ResourceBlock = {
  uri: string;
  mimeType?: string;
  text?: string;
  blob?: string;  // base64; not used in v1 but part of MCP contract
};

export type ListedResource = {
  uri: string;
  name: string;
  description?: string;
  mimeType?: string;
};

// ═══════════════════════════════════════════════════════════════════════
// `resources/list` — enumerate the common, useful-right-now resources
// so clients that don't know the URI scheme can still browse. Tier-1
// listing: scripts (up to 100) + spaces + recent conversations +
// discovered briefs. Templates handle the long tail.
// ═══════════════════════════════════════════════════════════════════════

const MAX_LIST = 200;

export async function listResources(universe: string): Promise<ListedResource[]> {
  const out: ListedResource[] = [];

  // Spaces as folder-like entries.
  for (const s of await listSpaces(universe)) {
    out.push({
      uri: buildUri({ kind: 'space', space: s.name }),
      name: `Space: ${s.name}`,
      description: `Overview of ${s.name} — services, top-level scripts, counts.`,
      mimeType: 'text/markdown',
    });
    if (out.length >= MAX_LIST) return out;
  }

  // Scripts — bundled source + summary is the highest-value pin the AI
  // can make, so we surface them early.
  const scripts = await listScripts(universe);
  for (const s of scripts) {
    const relPath = path.relative(
      path.join(universe, 'Spaces', s.space),
      s.folder,
    ).split(path.sep).join('/');
    out.push({
      uri: buildUri({ kind: 'script', space: s.space, relPath }),
      name: `Script: ${s.space}/${relPath}`,
      description: `${s.class} — source + summary`,
      mimeType: 'text/markdown',
    });
    if (out.length >= MAX_LIST) return out;
  }

  // Conversations — Workshop archive.
  const sessionDir = path.join(universe, '.eustress', 'knowledge', 'sessions');
  try {
    const files = await fs.readdir(sessionDir);
    const jsons = files.filter(f => f.endsWith('.json')).sort().reverse();
    for (const file of jsons.slice(0, 20)) {
      const id = file.slice(0, -'.json'.length);
      out.push({
        uri: buildUri({ kind: 'conversation', relPath: id }),
        name: `Workshop conversation: ${id}`,
        description: 'Persisted Workshop chat history',
        mimeType: 'application/json',
      });
      if (out.length >= MAX_LIST) return out;
    }
  } catch { /* no sessions yet, fine */ }

  // Briefs — `ideation_brief.toml` files anywhere in any Space.
  const briefs = await findBriefs(universe);
  for (const b of briefs) {
    out.push({
      uri: buildUri({ kind: 'brief', relPath: b.product }),
      name: `Brief: ${b.product}`,
      description: `Product ideation brief at ${b.relPath}`,
      mimeType: 'application/toml',
    });
    if (out.length >= MAX_LIST) return out;
  }

  return out;
}

// ═══════════════════════════════════════════════════════════════════════
// `resources/read` — resolve one URI → content block.
// ═══════════════════════════════════════════════════════════════════════

export async function readResource(
  universe: string,
  raw: string,
): Promise<ResourceBlock> {
  const uri = parseUri(raw);
  if (!uri) throw new Error(`Malformed Eustress URI: ${raw}`);

  switch (uri.kind) {
    case 'space':         return await readSpace(universe, uri);
    case 'script':        return await readScript(universe, uri);
    case 'entity':        return await readEntity(universe, uri);
    case 'file':          return await readFile(universe, uri);
    case 'conversation':  return await readConversation(universe, uri);
    case 'brief':         return await readBrief(universe, uri);
  }
}

// ─── space ─────────────────────────────────────────────────────────────

async function readSpace(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const space = uri.space!;
  const spaceDir = path.join(universe, 'Spaces', space);
  if (!fsSync.existsSync(spaceDir)) {
    throw new Error(`Space not found: ${space}`);
  }

  const services: string[] = [];
  try {
    const entries = await fs.readdir(spaceDir, { withFileTypes: true });
    for (const e of entries) {
      if (e.isDirectory() && !e.name.startsWith('.')) services.push(e.name);
    }
  } catch { /* empty Space — fine */ }

  const scripts = (await listScripts(universe, space)).filter(s => s.space === space);

  // Markdown overview — what the AI would want to see when pinning a Space.
  const parts: string[] = [];
  parts.push(`# Space: ${space}`);
  parts.push('');
  parts.push(`**Root:** \`${spaceDir}\``);
  parts.push('');
  parts.push(`## Services (${services.length})`);
  for (const s of services.sort()) parts.push(`- ${s}`);
  parts.push('');
  parts.push(`## Scripts (${scripts.length})`);
  for (const s of scripts) {
    const rel = path.relative(spaceDir, s.folder).split(path.sep).join('/');
    parts.push(`- [${s.class}] \`${rel}\``);
  }
  return {
    uri: uri.raw,
    mimeType: 'text/markdown',
    text: parts.join('\n'),
  };
}

// ─── script ────────────────────────────────────────────────────────────

async function readScript(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const space = uri.space!;
  const folder = path.join(universe, 'Spaces', space, uri.relPath);
  if (!fsSync.existsSync(folder)) {
    throw new Error(`Script folder not found: ${folder}`);
  }

  const name = path.basename(folder);
  const sourcePath = await findFirstExisting([
    path.join(folder, `${name}.rune`),
    path.join(folder, `${name}.luau`),
    path.join(folder, `${name}.soul`),
    path.join(folder, 'Source.rune'),
  ]);
  const summaryPath = await findFirstExisting([
    path.join(folder, `${name}.md`),
    path.join(folder, 'Summary.md'),
  ]);
  const instanceToml = path.join(folder, '_instance.toml');

  const parts: string[] = [];
  parts.push(`# Script: ${space}/${uri.relPath}`);
  parts.push('');
  parts.push(`**Folder:** \`${folder}\``);

  if (fsSync.existsSync(instanceToml)) {
    const toml = await fs.readFile(instanceToml, 'utf8');
    const klass = extractToml(toml, 'class_name') ?? '(unknown)';
    parts.push(`**Class:** ${klass}`);
  }
  parts.push('');

  if (summaryPath) {
    parts.push('## Summary');
    parts.push('');
    const { text, truncated } = await readCapped(summaryPath);
    parts.push(text.trimEnd());
    if (truncated) parts.push(`\n_[truncated at ${MAX_FILE_BYTES} bytes]_`);
    parts.push('');
  }

  if (sourcePath) {
    parts.push('## Source');
    parts.push('');
    const lang = sourcePath.endsWith('.luau') ? 'luau'
               : sourcePath.endsWith('.rune') ? 'rust'  // no hljs rune highlighter; rust is closest
               : '';
    parts.push('```' + lang);
    const { text, truncated } = await readCapped(sourcePath);
    parts.push(text.trimEnd());
    parts.push('```');
    if (truncated) parts.push(`\n_[source truncated at ${MAX_FILE_BYTES} bytes]_`);
  } else {
    parts.push('_(no source file found)_');
  }

  return {
    uri: uri.raw,
    mimeType: 'text/markdown',
    text: parts.join('\n'),
  };
}

// ─── entity ────────────────────────────────────────────────────────────

async function readEntity(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const space = uri.space!;
  const target = path.join(universe, 'Spaces', space, uri.relPath);

  // Entities can be folder-based (read `_instance.toml` inside) OR
  // flat-file (read the `.part.toml` / `.model.toml` / etc. directly).
  let tomlPath: string;
  if (fsSync.existsSync(target) && fsSync.statSync(target).isDirectory()) {
    tomlPath = path.join(target, '_instance.toml');
  } else {
    tomlPath = target;
  }

  if (!fsSync.existsSync(tomlPath)) {
    throw new Error(`Entity not found: ${target}`);
  }

  const { text, truncated } = await readCapped(tomlPath);
  const klass = extractToml(text, 'class_name') ?? '(unknown)';
  const name  = extractToml(text, 'name') ?? path.basename(path.dirname(tomlPath));

  const parts: string[] = [];
  parts.push(`# Entity: ${name}`);
  parts.push('');
  parts.push(`**Class:** ${klass}`);
  parts.push(`**Path:** \`${tomlPath}\``);
  parts.push('');
  parts.push('## _instance.toml');
  parts.push('');
  parts.push('```toml');
  parts.push(text.trimEnd());
  parts.push('```');
  if (truncated) parts.push(`\n_[truncated at ${MAX_FILE_BYTES} bytes]_`);
  return {
    uri: uri.raw,
    mimeType: 'text/markdown',
    text: parts.join('\n'),
  };
}

// ─── file ──────────────────────────────────────────────────────────────

async function readFile(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const space = uri.space!;
  const target = path.join(universe, 'Spaces', space, uri.relPath);
  if (!fsSync.existsSync(target)) {
    throw new Error(`File not found: ${target}`);
  }

  // Minimal binary-file rejection — MCP clients expect text resources.
  // A production system would sniff more thoroughly (e.g. libmagic).
  const ext = path.extname(target).toLowerCase();
  const binaryExt = new Set(['.png', '.jpg', '.jpeg', '.gif', '.webp', '.mp3',
    '.wav', '.ogg', '.glb', '.gltf', '.fbx', '.obj', '.stl', '.ktx2']);
  if (binaryExt.has(ext)) {
    throw new Error(`Binary file not exposed as text resource: ${target}`);
  }

  const { text, truncated } = await readCapped(target);
  const mimeType = ext === '.md' ? 'text/markdown'
                 : ext === '.toml' ? 'application/toml'
                 : ext === '.json' ? 'application/json'
                 : 'text/plain';

  return {
    uri: uri.raw,
    mimeType,
    text: truncated ? text + `\n\n[truncated at ${MAX_FILE_BYTES} bytes]` : text,
  };
}

// ─── conversation ──────────────────────────────────────────────────────

async function readConversation(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const id = uri.relPath;
  const sessionFile = path.join(
    universe, '.eustress', 'knowledge', 'sessions', `${id}.json`,
  );
  if (!fsSync.existsSync(sessionFile)) {
    throw new Error(`No such conversation: ${id}`);
  }
  const text = await fs.readFile(sessionFile, 'utf8');
  return { uri: uri.raw, mimeType: 'application/json', text };
}

// ─── brief ─────────────────────────────────────────────────────────────

async function readBrief(universe: string, uri: EustressUri): Promise<ResourceBlock> {
  const product = uri.relPath;
  const briefs = await findBriefs(universe);
  const match = briefs.find(b => b.product === product);
  if (!match) throw new Error(`No ideation brief for product: ${product}`);
  const { text, truncated } = await readCapped(match.path);
  return {
    uri: uri.raw,
    mimeType: 'application/toml',
    text: truncated ? text + `\n\n# [truncated at ${MAX_FILE_BYTES} bytes]` : text,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Reverse mapping — filesystem path → URI. Used by the watcher so a
// change to `<universe>/Spaces/Space1/SoulService/foo/foo.rune` emits
// `resources/updated` for `eustress://script/Space1/SoulService/foo`.
// ═══════════════════════════════════════════════════════════════════════

export function pathToUri(universe: string, absPath: string): string | null {
  const rel = path.relative(universe, absPath).split(path.sep).join('/');

  // .eustress/knowledge/sessions/<id>.json → conversation
  const sessionMatch = rel.match(/^\.eustress\/knowledge\/sessions\/(.+)\.json$/);
  if (sessionMatch) {
    return buildUri({ kind: 'conversation', relPath: sessionMatch[1] });
  }

  // Spaces/<space>/.../ideation_brief.toml → brief
  const briefMatch = rel.match(/^Spaces\/([^\/]+)\/(.+)\/ideation_brief\.toml$/);
  if (briefMatch) {
    // Product name = parent folder of ideation_brief.toml.
    const product = path.basename(briefMatch[2]);
    return buildUri({ kind: 'brief', relPath: product });
  }

  // Spaces/<space>/<relpath>
  const spaceMatch = rel.match(/^Spaces\/([^\/]+)\/(.+)$/);
  if (!spaceMatch) return null;
  const space = spaceMatch[1];
  const innerRel = spaceMatch[2];

  // script: innerRel ends at a folder that contains `<folder>.<ext>` or
  // `Source.<ext>`. Rather than re-deriving, check if the PARENT folder
  // of the changed file is a script folder.
  const fileName = path.basename(innerRel);
  const parentRel = path.dirname(innerRel);
  if (parentRel && parentRel !== '.') {
    const parentAbs = path.join(universe, 'Spaces', space, parentRel);
    const parentName = path.basename(parentRel);
    const isScriptFile = fileName === `${parentName}.rune`
      || fileName === `${parentName}.md`
      || fileName === 'Source.rune'
      || fileName === 'Summary.md'
      || fileName === '_instance.toml';
    if (isScriptFile && fsSync.existsSync(path.join(parentAbs, '_instance.toml'))) {
      return buildUri({ kind: 'script', space, relPath: parentRel.split(path.sep).join('/') });
    }
  }

  // entity: the file itself is `_instance.toml` inside a folder.
  if (fileName === '_instance.toml' && parentRel && parentRel !== '.') {
    return buildUri({ kind: 'entity', space, relPath: parentRel.split(path.sep).join('/') });
  }

  // Everything else that's a text-ish file is a generic `file` resource.
  return buildUri({ kind: 'file', space, relPath: innerRel.split(path.sep).join('/') });
}

// ═══════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════

async function findFirstExisting(candidates: string[]): Promise<string | null> {
  for (const c of candidates) {
    if (fsSync.existsSync(c)) return c;
  }
  return null;
}

type BriefEntry = { product: string; path: string; relPath: string };

async function findBriefs(universe: string): Promise<BriefEntry[]> {
  const out: BriefEntry[] = [];
  const spacesDir = path.join(universe, 'Spaces');
  if (!fsSync.existsSync(spacesDir)) return out;

  // Shallow-ish walk. `ideation_brief.toml` files live under
  // `Spaces/*/Workspace/<product>/` or similar; up to 4 levels deep
  // covers every layout Workshop produces.
  await walkForBriefs(spacesDir, universe, 4, out);
  return out;
}

async function walkForBriefs(
  dir: string, universe: string, depth: number, out: BriefEntry[],
): Promise<void> {
  if (depth < 0 || out.length >= 100) return;
  let entries: import('node:fs').Dirent[] = [];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch { return; }

  for (const e of entries) {
    if (e.isFile() && e.name === 'ideation_brief.toml') {
      out.push({
        product: path.basename(dir),
        path: path.join(dir, e.name),
        relPath: path.relative(universe, path.join(dir, e.name)),
      });
      continue;
    }
    if (e.isDirectory() && !e.name.startsWith('.')) {
      await walkForBriefs(path.join(dir, e.name), universe, depth - 1, out);
    }
  }
}

File: eustress/crates/mcp-server/src/resources.rs
// Resource resolvers — one per URI kind. Each resolver reads what it needs
// from the Universe filesystem and returns a `{uri, mimeType, text}` block
// fit for `resources/read`.
//
// Tools are actions the AI invokes; resources are memory the AI pins and
// refers back to. That separation means `eustress_read_script` (tool) and
// `eustress://script/...` (resource) can format the same data differently
// — JSON for programmatic consumption vs. markdown for pinning.

use serde::Serialize;
use std::path::{Path, PathBuf};

use crate::uri::{self, EustressUri, UriKind};
use crate::universe::{
    extract_toml, list_scripts, list_spaces, read_capped, MAX_FILE_BYTES,
};

#[derive(Serialize, Debug, Clone)]
pub struct ResourceBlock {
    pub uri: String,
    #[serde(rename = "mimeType", skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blob: Option<String>,
}

#[derive(Serialize, Debug, Clone)]
pub struct ListedResource {
    pub uri: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(rename = "mimeType", skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
}

const MAX_LIST: usize = 200;

pub fn list_resources(universe: &Path) -> Vec<ListedResource> {
    let mut out: Vec<ListedResource> = Vec::new();

    for s in list_spaces(universe) {
        out.push(ListedResource {
            uri: uri::build(UriKind::Space, Some(&s.name), ""),
            name: format!("Space: {}", s.name),
            description: Some(format!(
                "Overview of {} — services, top-level scripts, counts.",
                s.name
            )),
            mime_type: Some("text/markdown".into()),
        });
        if out.len() >= MAX_LIST {
            return out;
        }
    }

    let scripts = list_scripts(universe, None);
    for s in &scripts {
        let space_root = universe.join("Spaces").join(&s.space);
        let rel = match Path::new(&s.folder).strip_prefix(&space_root) {
            Ok(r) => path_slash(r),
            Err(_) => continue,
        };
        out.push(ListedResource {
            uri: uri::build(UriKind::Script, Some(&s.space), &rel),
            name: format!("Script: {}/{}", s.space, rel),
            description: Some(format!("{} — source + summary", s.class)),
            mime_type: Some("text/markdown".into()),
        });
        if out.len() >= MAX_LIST {
            return out;
        }
    }

    // Conversations — Workshop archive.
    let session_dir = universe.join(".eustress").join("knowledge").join("sessions");
    if let Ok(read) = std::fs::read_dir(&session_dir) {
        let mut sessions: Vec<String> = read
            .flatten()
            .filter_map(|e| {
                let name = e.file_name().to_string_lossy().to_string();
                name.strip_suffix(".json").map(|s| s.to_string())
            })
            .collect();
        sessions.sort();
        sessions.reverse();
        for id in sessions.into_iter().take(20) {
            out.push(ListedResource {
                uri: uri::build(UriKind::Conversation, None, &id),
                name: format!("Workshop conversation: {id}"),
                description: Some("Persisted Workshop chat history".into()),
                mime_type: Some("application/json".into()),
            });
            if out.len() >= MAX_LIST {
                return out;
            }
        }
    }

    for b in find_briefs(universe) {
        out.push(ListedResource {
            uri: uri::build(UriKind::Brief, None, &b.product),
            name: format!("Brief: {}", b.product),
            description: Some(format!("Product ideation brief at {}", b.rel_path)),
            mime_type: Some("application/toml".into()),
        });
        if out.len() >= MAX_LIST {
            return out;
        }
    }

    out
}

pub fn read_resource(universe: &Path, raw: &str) -> Result<ResourceBlock, String> {
    let u = uri::parse(raw).ok_or_else(|| format!("Malformed Eustress URI: {raw}"))?;
    match u.kind {
        UriKind::Space => read_space(universe, &u),
        UriKind::Script => read_script(universe, &u),
        UriKind::Entity => read_entity(universe, &u),
        UriKind::File => read_file(universe, &u),
        UriKind::Conversation => read_conversation(universe, &u),
        UriKind::Brief => read_brief(universe, &u),
    }
}

fn read_space(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let space = u.space.as_deref().ok_or("space URI missing name")?;
    let space_dir = universe.join("Spaces").join(space);
    if !space_dir.is_dir() {
        return Err(format!("Space not found: {space}"));
    }

    let mut services: Vec<String> = Vec::new();
    if let Ok(read) = std::fs::read_dir(&space_dir) {
        for e in read.flatten() {
            if e.file_type().map(|ft| ft.is_dir()).unwrap_or(false) {
                let name = e.file_name().to_string_lossy().to_string();
                if !name.starts_with('.') {
                    services.push(name);
                }
            }
        }
    }
    services.sort();

    let scripts = list_scripts(universe, Some(space));

    let mut parts = Vec::new();
    parts.push(format!("# Space: {space}"));
    parts.push(String::new());
    parts.push(format!("**Root:** `{}`", space_dir.display()));
    parts.push(String::new());
    parts.push(format!("## Services ({})", services.len()));
    for s in &services {
        parts.push(format!("- {s}"));
    }
    parts.push(String::new());
    parts.push(format!("## Scripts ({})", scripts.len()));
    for s in &scripts {
        let rel = Path::new(&s.folder)
            .strip_prefix(&space_dir)
            .map(path_slash)
            .unwrap_or_else(|_| s.folder.clone());
        parts.push(format!("- [{}] `{}`", s.class, rel));
    }

    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some("text/markdown".into()),
        text: Some(parts.join("\n")),
        blob: None,
    })
}

fn read_script(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let space = u.space.as_deref().ok_or("script URI missing space")?;
    let folder = universe.join("Spaces").join(space).join(&u.rel_path);
    if !folder.is_dir() {
        return Err(format!("Script folder not found: {}", folder.display()));
    }

    let name = folder.file_name().map(|s| s.to_string_lossy().to_string()).unwrap_or_default();
    let source_path = find_first_existing(&[
        folder.join(format!("{name}.rune")),
        folder.join(format!("{name}.luau")),
        folder.join(format!("{name}.soul")),
        folder.join("Source.rune"),
    ]);
    let summary_path = find_first_existing(&[
        folder.join(format!("{name}.md")),
        folder.join("Summary.md"),
    ]);
    let instance_toml = folder.join("_instance.toml");

    let mut parts: Vec<String> = Vec::new();
    parts.push(format!("# Script: {space}/{}", u.rel_path));
    parts.push(String::new());
    parts.push(format!("**Folder:** `{}`", folder.display()));

    if instance_toml.exists() {
        if let Ok(text) = std::fs::read_to_string(&instance_toml) {
            let klass = extract_toml(&text, "class_name").unwrap_or_else(|| "(unknown)".into());
            parts.push(format!("**Class:** {klass}"));
        }
    }
    parts.push(String::new());

    if let Some(sp) = &summary_path {
        parts.push("## Summary".into());
        parts.push(String::new());
        let r = read_capped(sp).map_err(|e| e.to_string())?;
        parts.push(r.text.trim_end().to_string());
        if r.truncated {
            parts.push(format!("\n_[truncated at {MAX_FILE_BYTES} bytes]_"));
        }
        parts.push(String::new());
    }

    if let Some(src) = &source_path {
        parts.push("## Source".into());
        parts.push(String::new());
        let lang = if src.extension().and_then(|s| s.to_str()) == Some("luau") {
            "luau"
        } else if src.extension().and_then(|s| s.to_str()) == Some("rune") {
            "rust" // no hljs rune highlighter; rust is closest
        } else {
            ""
        };
        parts.push(format!("```{lang}"));
        let r = read_capped(src).map_err(|e| e.to_string())?;
        parts.push(r.text.trim_end().to_string());
        parts.push("```".into());
        if r.truncated {
            parts.push(format!("\n_[source truncated at {MAX_FILE_BYTES} bytes]_"));
        }
    } else {
        parts.push("_(no source file found)_".into());
    }

    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some("text/markdown".into()),
        text: Some(parts.join("\n")),
        blob: None,
    })
}

fn read_entity(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let space = u.space.as_deref().ok_or("entity URI missing space")?;
    let target = universe.join("Spaces").join(space).join(&u.rel_path);

    let toml_path = if target.is_dir() {
        target.join("_instance.toml")
    } else {
        target.clone()
    };
    if !toml_path.exists() {
        return Err(format!("Entity not found: {}", target.display()));
    }

    let r = read_capped(&toml_path).map_err(|e| e.to_string())?;
    let klass = extract_toml(&r.text, "class_name").unwrap_or_else(|| "(unknown)".into());
    let fallback_name = toml_path
        .parent()
        .and_then(|p| p.file_name())
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();
    let name = extract_toml(&r.text, "name").unwrap_or(fallback_name);

    let mut parts = vec![
        format!("# Entity: {name}"),
        String::new(),
        format!("**Class:** {klass}"),
        format!("**Path:** `{}`", toml_path.display()),
        String::new(),
        "## _instance.toml".into(),
        String::new(),
        "```toml".into(),
        r.text.trim_end().to_string(),
        "```".into(),
    ];
    if r.truncated {
        parts.push(format!("\n_[truncated at {MAX_FILE_BYTES} bytes]_"));
    }

    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some("text/markdown".into()),
        text: Some(parts.join("\n")),
        blob: None,
    })
}

fn read_file(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let space = u.space.as_deref().ok_or("file URI missing space")?;
    let target = universe.join("Spaces").join(space).join(&u.rel_path);
    if !target.exists() {
        return Err(format!("File not found: {}", target.display()));
    }

    let ext = target
        .extension()
        .and_then(|s| s.to_str())
        .map(|s| format!(".{}", s.to_lowercase()))
        .unwrap_or_default();
    const BINARY: &[&str] = &[
        ".png", ".jpg", ".jpeg", ".gif", ".webp", ".mp3", ".wav", ".ogg",
        ".glb", ".gltf", ".fbx", ".obj", ".stl", ".ktx2",
    ];
    if BINARY.contains(&ext.as_str()) {
        return Err(format!(
            "Binary file not exposed as text resource: {}",
            target.display()
        ));
    }

    let r = read_capped(&target).map_err(|e| e.to_string())?;
    let mime = match ext.as_str() {
        ".md" => "text/markdown",
        ".toml" => "application/toml",
        ".json" => "application/json",
        _ => "text/plain",
    };
    let text = if r.truncated {
        format!("{}\n\n[truncated at {MAX_FILE_BYTES} bytes]", r.text)
    } else {
        r.text
    };

    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some(mime.into()),
        text: Some(text),
        blob: None,
    })
}

fn read_conversation(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let id = &u.rel_path;
    let session_file = universe
        .join(".eustress")
        .join("knowledge")
        .join("sessions")
        .join(format!("{id}.json"));
    if !session_file.exists() {
        return Err(format!("No such conversation: {id}"));
    }
    let text = std::fs::read_to_string(&session_file).map_err(|e| e.to_string())?;
    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some("application/json".into()),
        text: Some(text),
        blob: None,
    })
}

fn read_brief(universe: &Path, u: &EustressUri) -> Result<ResourceBlock, String> {
    let product = &u.rel_path;
    let briefs = find_briefs(universe);
    let m = briefs
        .iter()
        .find(|b| b.product == *product)
        .ok_or_else(|| format!("No ideation brief for product: {product}"))?;
    let r = read_capped(&m.path).map_err(|e| e.to_string())?;
    let text = if r.truncated {
        format!("{}\n\n# [truncated at {MAX_FILE_BYTES} bytes]", r.text)
    } else {
        r.text
    };
    Ok(ResourceBlock {
        uri: u.raw.clone(),
        mime_type: Some("application/toml".into()),
        text: Some(text),
        blob: None,
    })
}

/// Reverse mapping — filesystem path → URI. Used by the watcher so a change
/// to `<universe>/Spaces/Space1/SoulService/foo/foo.rune` emits
/// `resources/updated` for `eustress://script/Space1/SoulService/foo`.
pub fn path_to_uri(universe: &Path, abs_path: &Path) -> Option<String> {
    let rel = abs_path.strip_prefix(universe).ok()?;
    let rel_str = path_slash(rel);
    let parts: Vec<&str> = rel_str.split('/').collect();

    // `.eustress/knowledge/sessions/<id>.json` → conversation
    if parts.len() == 4
        && parts[0] == ".eustress"
        && parts[1] == "knowledge"
        && parts[2] == "sessions"
    {
        if let Some(id) = parts[3].strip_suffix(".json") {
            return Some(uri::build(UriKind::Conversation, None, id));
        }
    }

    // `Spaces/<space>/.../ideation_brief.toml` → brief
    if parts.len() >= 3 && parts[0] == "Spaces" {
        if let Some(last) = parts.last() {
            if *last == "ideation_brief.toml" && parts.len() >= 4 {
                let product = parts[parts.len() - 2];
                return Some(uri::build(UriKind::Brief, None, product));
            }
        }
    }

    // `Spaces/<space>/<inner...>`
    if parts.len() < 3 || parts[0] != "Spaces" {
        return None;
    }
    let space = parts[1];
    let inner = parts[2..].join("/");
    let file_name = parts.last().copied().unwrap_or("");
    let parent_parts = &parts[2..parts.len() - 1];

    if !parent_parts.is_empty() {
        let parent_abs = universe.join("Spaces").join(space).join(parent_parts.join("/"));
        let parent_name = parent_parts.last().copied().unwrap_or("");
        let is_script_file = file_name == format!("{parent_name}.rune")
            || file_name == format!("{parent_name}.md")
            || file_name == "Source.rune"
            || file_name == "Summary.md"
            || file_name == "_instance.toml";
        if is_script_file && parent_abs.join("_instance.toml").exists() {
            return Some(uri::build(
                UriKind::Script,
                Some(space),
                &parent_parts.join("/"),
            ));
        }
    }

    if file_name == "_instance.toml" && !parent_parts.is_empty() {
        return Some(uri::build(
            UriKind::Entity,
            Some(space),
            &parent_parts.join("/"),
        ));
    }

    Some(uri::build(UriKind::File, Some(space), &inner))
}

fn find_first_existing(candidates: &[PathBuf]) -> Option<PathBuf> {
    candidates.iter().find(|p| p.exists()).cloned()
}

pub struct BriefEntry {
    pub product: String,
    pub path: PathBuf,
    pub rel_path: String,
}

pub fn find_briefs(universe: &Path) -> Vec<BriefEntry> {
    let mut out = Vec::new();
    let spaces_dir = universe.join("Spaces");
    if !spaces_dir.is_dir() {
        return out;
    }
    walk_for_briefs(&spaces_dir, universe, 4, &mut out);
    out
}

fn walk_for_briefs(dir: &Path, universe: &Path, depth: i32, out: &mut Vec<BriefEntry>) {
    if depth < 0 || out.len() >= 100 {
        return;
    }
    let read = match std::fs::read_dir(dir) {
        Ok(r) => r,
        Err(_) => return,
    };
    for e in read.flatten() {
        let ft = match e.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        let name = e.file_name().to_string_lossy().to_string();
        if ft.is_file() && name == "ideation_brief.toml" {
            let product = dir
                .file_name()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_default();
            let path = e.path();
            let rel_path = path
                .strip_prefix(universe)
                .map(path_slash)
                .unwrap_or_default();
            out.push(BriefEntry { product, path, rel_path });
            continue;
        }
        if ft.is_dir() && !name.starts_with('.') {
            walk_for_briefs(&e.path(), universe, depth - 1, out);
        }
    }
}

fn path_slash(p: &Path) -> String {
    p.components()
        .map(|c| c.as_os_str().to_string_lossy().into_owned())
        .collect::<Vec<_>>()
        .join("/")
}

